realm/realm-library/src/main/cpp/objectserver_shared.hpp

```diff
@@ -19,123 +19,106 @@
-#include "jni_util/java_global_weak_ref.hpp"
-#include "jni_util/java_method.hpp"
 
+using namespace realm;
 
-// Wrapper class for realm::Session. This allows us to manage the C++ session and callback lifecycle correctly.
-// TODO Use OS SyncSession instead
-class JniSession {
+// Wrapper class for SyncConfig. This is required as we need to keep track of both the Java session
+// object as part of the configuration.
+// NOTICE: It is an requirement that the Java `SyncSession` object is created before this and
+// is only GC'ed after this object has been destroyed.
+class JniConfigWrapper {
 
 public:
-    JniSession(const JniSession&) = delete;
-    JniSession& operator=(const JniSession&) = delete;
-    JniSession(JniSession&&) = delete;
-    JniSession& operator=(JniSession&&) = delete;
-
-    JniSession(JNIEnv* env, std::string local_realm_path, jobject java_session_obj)
-    : m_java_session_ref(std::make_shared<realm::jni_util::JavaGlobalWeakRef>(env, java_session_obj))
+    JniConfigWrapper(const JniConfigWrapper&) = delete;
+
+    JniConfigWrapper& operator=(const JniConfigWrapper&) = delete;
+    JniConfigWrapper(JniConfigWrapper&&) = delete;
+    JniConfigWrapper& operator=(JniConfigWrapper&&) = delete;
+
+    // Non-sync constructor
+    JniConfigWrapper(JNIEnv*, Realm::Config* config) : m_config (config) {}
+
+    // Sync constructor
+    JniConfigWrapper(REALM_UNUSED JNIEnv* env,
+                     REALM_UNUSED Realm::Config* config,
+                     REALM_UNUSED jstring sync_realm_url,
+                     REALM_UNUSED jstring sync_realm_auth_url,
+                     REALM_UNUSED jstring sync_user_identity,
+                     REALM_UNUSED jstring sync_refresh_token)
     {
-        extern std::unique_ptr<realm::sync::Client> sync_client;
-        // Get the coordinator for the given path, or null if there is none
-        m_sync_session = new realm::sync::Session(*sync_client, local_realm_path);
-        // error_handler could be called after JniSession destructed. So we need to pass a weak ref to lambda to avoid
-        // the corrupted pointer.
-        std::weak_ptr<realm::jni_util::JavaGlobalWeakRef> weak_session_ref(m_java_session_ref);
-        auto sync_transact_callback = [local_realm_path](realm::VersionID, realm::VersionID) {
-            auto coordinator = realm::_impl::RealmCoordinator::get_existing_coordinator(
-                    realm::StringData(local_realm_path));
-            if (coordinator) {
-                coordinator->wake_up_notifier_worker();
-            }
+#ifdef REALM_ENABLE_SYNC
+        m_config = config;
+
+        // error handler will be called form the sync client thread
+        auto error_handler = [=](std::shared_ptr<SyncSession> session, SyncError error) {
+            realm::jni_util::Log::d("error_handler lambda invoked");
+
+            JNIEnv *env = realm::jni_util::JniUtils::get_env(true);
+            env->CallStaticVoidMethod(java_syncmanager,
```

Comment:
This will propagate the error to Java via `SyncSession.notifySessionError`. The default error handler is a no-op, however if the user specify a custom error handler that may throw, then the caller thread will crash causing probably the sync client thread to crash as well